server/src/main/java/org/elasticsearch/search/profile/query/ProfileCollectorManager.java

```diff
@@ -8,54 +8,112 @@
 
 package org.elasticsearch.search.profile.query;
 
+import org.apache.lucene.sandbox.search.ProfilerCollector;
 import org.apache.lucene.search.Collector;
 import org.apache.lucene.search.CollectorManager;
 
 import java.io.IOException;
+import java.util.ArrayList;
 import java.util.Collection;
-import java.util.Collections;
 import java.util.List;
-import java.util.stream.Collectors;
 
 /**
  * A {@link CollectorManager} that takes another CollectorManager as input and wraps all Collectors generated by it
  * in an {@link InternalProfileCollector}. It delegates all the profiling to the generated collectors via {@link #getCollectorTree()}
- * and joins them up when its {@link #reduce} method is called. The profile result can
+ * and joins the different collector trees together when its {@link #reduce} method is called.
+ * Supports optionally providing sub-collector managers for top docs as well as aggs collection, so that each
+ * {@link InternalProfileCollector} created is provided with the corresponding sub-collectors that are children of the top-level collector.
+ * @param <T> the return type of the wrapped collector manager, which the reduce method returns.
  */
 public final class ProfileCollectorManager<T> implements CollectorManager<InternalProfileCollector, T> {
 
-    private final CollectorManager<Collector, T> collectorManager;
+    private final CollectorManager<? extends Collector, T> collectorManager;
     private final String reason;
+    private final ProfileCollectorManager<?> topDocsSubCollectorManager;
+    private final ProfileCollectorManager<?> aggsSubCollectorManager;
+    // this is a bit of a hack: it allows us to retrieve the last collector that newCollector has returned for sub-collector managers,
+    // so that we can provide them to InternalProfileCollector's constructor as children.
+    private InternalProfileCollector profileCollector;
+
     private CollectorResult collectorTree;
 
-    @SuppressWarnings("unchecked")
     public ProfileCollectorManager(CollectorManager<? extends Collector, T> collectorManager, String reason) {
-        this.collectorManager = (CollectorManager<Collector, T>) collectorManager;
+        this(collectorManager, reason, null, null);
+    }
+
+    public ProfileCollectorManager(
+        CollectorManager<? extends Collector, T> collectorManager,
```

Comment:
Is it save to assume that in this case the parent collector manager needs to be a MultiCollectorManager? Can we restrict this constructor to this case?

Reply:
No, it can either be the collector manager that aggs use, or top docs that we use in the dfs phase, or the collector manager (yet to be created) for query phase collector.